test/runtime/cartesian_vector_test.cpp

```diff
@@ -92,352 +79,170 @@ TEST_CASE("cartesian_vector operations", "[vector]")
   {
     cartesian_vector v1{1.0, 2.0, 3.0};
     cartesian_vector v2{1, 2, 3};
     cartesian_vector v3{1.1, 2.0, 3.0};
-    cartesian_vector v4{1.0, 2.1, 3.0};
-    cartesian_vector v5{1.0, 2.0, 3.1};
     REQUIRE(v1 == v2);
     REQUIRE(v1 != v3);
-    REQUIRE(v1 != v4);
-    REQUIRE(v1 != v5);
   }
 
-  SECTION("cartesian_vector scalar product")
+  SECTION("scalar_product and vector_product")
   {
-    SECTION("double * double")
-    {
-      cartesian_vector v1{1.0, 2.0, 3.0};
-      cartesian_vector v2{4.0, 5.0, 6.0};
-      REQUIRE(scalar_product(v1, v2) == 32.0);
-    }
-
-    SECTION("double * int")
-    {
-      cartesian_vector v1{1.0, 2.0, 3.0};
-      cartesian_vector v2{4, 5, 6};
-      REQUIRE(scalar_product(v1, v2) == 32.0);
-    }
-
-    SECTION("int * double")
-    {
-      cartesian_vector v1{1, 2, 3};
-      cartesian_vector v2{4.0, 5.0, 6.0};
-      REQUIRE(scalar_product(v1, v2) == 32.0);
-    }
-
-    SECTION("int * int")
-    {
-      cartesian_vector v1{1, 2, 3};
-      cartesian_vector v2{4, 5, 6};
-      REQUIRE(scalar_product(v1, v2) == 32);
-    }
+    cartesian_vector a{1, 2, 3};
+    cartesian_vector b{4, 5, 6};
+    REQUIRE(scalar_product(a, b) == 32);
+    auto r = vector_product(a, b);
+    REQUIRE(r[0] == -3);
+    REQUIRE(r[1] == 6);
+    REQUIRE(r[2] == -3);
   }
 
-  SECTION("cartesian_vector vector product")
+  SECTION("magnitude and unit (floating types)")
   {
-    SECTION("double * double")
-    {
-      cartesian_vector v1{1.0, 2.0, 3.0};
-      cartesian_vector v2{4.0, 5.0, 6.0};
-      cartesian_vector result = vector_product(v1, v2);
-      REQUIRE(result[0] == -3.0);
-      REQUIRE(result[1] == 6.0);
-      REQUIRE(result[2] == -3.0);
-    }
-
-    SECTION("double * int")
-    {
-      cartesian_vector v1{1.0, 2.0, 3.0};
-      cartesian_vector v2{4, 5, 6};
-      cartesian_vector result = vector_product(v1, v2);
-      REQUIRE(result[0] == -3.0);
-      REQUIRE(result[1] == 6.0);
-      REQUIRE(result[2] == -3.0);
-    }
-
-    SECTION("int * double")
-    {
-      cartesian_vector v1{1, 2, 3};
-      cartesian_vector v2{4.0, 5.0, 6.0};
-      cartesian_vector result = vector_product(v1, v2);
-      REQUIRE(result[0] == -3.0);
-      REQUIRE(result[1] == 6.0);
-      REQUIRE(result[2] == -3.0);
-    }
-
-    SECTION("int * int")
-    {
-      cartesian_vector v1{1, 2, 3};
-      cartesian_vector v2{4, 5, 6};
-      cartesian_vector result = vector_product(v1, v2);
-      REQUIRE(result[0] == -3);
-      REQUIRE(result[1] == 6);
-      REQUIRE(result[2] == -3);
-    }
+    cartesian_vector<double> v{3.0, 4.0, 0.0};
+    REQUIRE_THAT(v.magnitude(), WithinAbs(5.0, 1e-12));
+    auto u = v.unit();
+    REQUIRE_THAT(u.magnitude(), WithinAbs(1.0, 1e-12));
+    REQUIRE_THAT(u[0], WithinAbs(3.0 / 5.0, 1e-12));
+    REQUIRE_THAT(u[1], WithinAbs(4.0 / 5.0, 1e-12));
+    REQUIRE_THAT(u[2], WithinAbs(0.0, 1e-12));
   }
-}
-
-TEST_CASE("cartesian_vector text output", "[vector][fmt][ostream]")
```

Comment:
Why did you remove my tests that were testing each text output separately for different representation types?